File: apps/desktop/src/main/model-audit.test.ts
import { describe, expect, it, vi } from "vitest";
import type {
  ModelAuditEvent,
  ModelAuditFinishInput,
  ModelAuditStartInput,
  ModelInvocation,
} from "@bubu/contracts";
import { buildModelAuditStart, generateAuditedModel, type AuditedModelRuntime } from "./model-audit.js";

const invocation: ModelInvocation = {
  provider: {
    id: "a".repeat(32), name: "Provider", kind: "openai",
    baseUrl: "https://user:secret@api.example.com/v1/?key=hidden", model: "model",
  },
  credential: "secret",
  system: "Plan safely",
  user: JSON.stringify({ question: "total", context: "synthetic only" }),
  maxOutputTokens: 100,
};

const context = {
  datasetId: "b".repeat(32),
  versionId: "c".repeat(32),
  disclosure: "schema-synthetic" as const,
  columns: [{ name: "Amount", type: "real" as const, nullable: false, unique: false }],
  syntheticRows: [[101.25], [102.25], [103.25]],
};

describe("audited model invocation", () => {
  it("summarizes and hashes the exact prompt without persisting it or URL secrets", () => {
    const audit = buildModelAuditStart(invocation, {
      purpose: "query-plan",
      target: { kind: "dataset", id: context.datasetId },
      contexts: [context],
      relationshipCount: 0,
    });
    expect(audit).toMatchObject({
      endpointOrigin: "https://api.example.com",
      datasetCount: 1,
      columnCount: 1,
      syntheticRowCount: 3,
      containsRawRows: false,
    });
    expect(JSON.stringify(audit)).not.toContain("total");
    expect(JSON.stringify(audit)).not.toContain("secret");
  });

  it("starts before provider I/O and persists terminal usage", async () => {
    const order: string[] = [];
    const started: ModelAuditEvent = {
      ...buildModelAuditStart(invocation, {
        purpose: "query-plan", target: { kind: "dataset", id: context.datasetId },
        contexts: [context], relationshipCount: 0,
      }),
      id: "d".repeat(32), status: "started", inputTokens: null, outputTokens: null,
      totalTokens: null, outputBytes: null, error: null,
      startedAt: "2026-07-17T00:00:00Z", finishedAt: null,
    };
    const finish = vi.fn(async (input: ModelAuditFinishInput) => ({ ...started, ...input } as ModelAuditEvent));
    const runtime: AuditedModelRuntime = {
      startModelAudit: async (_input: ModelAuditStartInput) => { order.push("audit"); return started; },
      generateModel: async () => {
        order.push("provider");
        return { providerId: invocation.provider.id, providerKind: "openai", model: "model", text: "{}", usage: { inputTokens: 10, outputTokens: 2, totalTokens: 12 } };
      },
      finishModelAudit: finish,
    };
    await generateAuditedModel(runtime, invocation, {
      purpose: "query-plan", target: { kind: "dataset", id: context.datasetId },
      contexts: [context], relationshipCount: 0,
    });
    expect(order).toEqual(["audit", "provider"]);
    expect(finish).toHaveBeenCalledWith(expect.objectContaining({ status: "succeeded", totalTokens: 12 }));
  });

  it("persists cancellation instead of leaving a started request", async () => {
    const auditInput = buildModelAuditStart(invocation, {
      purpose: "query-plan", target: { kind: "dataset", id: context.datasetId },
      contexts: [context], relationshipCount: 0,
    });
    const started = {
      ...auditInput, id: "e".repeat(32), status: "started" as const,
      inputTokens: null, outputTokens: null, totalTokens: null, outputBytes: null,
      error: null, startedAt: "2026-07-17T00:00:00Z", finishedAt: null,
    };
    const finish = vi.fn(async (input: ModelAuditFinishInput) => ({ ...started, ...input } as ModelAuditEvent));
    const controller = new AbortController();
    controller.abort();
    const runtime: AuditedModelRuntime = {
      startModelAudit: async () => started,
      generateModel: async () => { throw new Error("Operation cancelled"); },
      finishModelAudit: finish,
    };
    await expect(generateAuditedModel(runtime, invocation, {
      purpose: "query-plan", target: { kind: "dataset", id: context.datasetId },
      contexts: [context], relationshipCount: 0,
    }, controller.signal)).rejects.toThrow("cancelled");
    expect(finish).toHaveBeenCalledWith(expect.objectContaining({ status: "cancelled" }));
  });
});

File: services/data-core/internal/data/model_audit_test.go
package data

import (
	"context"
	"testing"
)

func datasetModelAuditInput(dataset DatasetSummary) ModelAuditStartInput {
	return ModelAuditStartInput{
		Purpose: "query-plan", Target: ModelAuditTarget{Kind: "dataset", ID: dataset.ID},
		Disclosure: "schema-synthetic", ProviderID: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
		ProviderKind: "openai", ProviderName: "Audited provider", Model: "model-example",
		EndpointOrigin: "https://api.example.com", DatasetCount: 1, ColumnCount: dataset.ColumnCount,
		SyntheticRowCount: 3, RelationshipCount: 0, PayloadBytes: 2_048,
		EstimatedInputTokens: 683, MaximumOutputTokens: 4_096,
		PayloadSHA256:   "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
		ContainsRawRows: false,
	}
}

func TestModelAuditPersistsDisclosureAndUsageWithoutPayload(t *testing.T) {
	service, dataset := importQueryFixture(t)
	event, err := service.StartModelAudit(context.Background(), datasetModelAuditInput(dataset))
	if err != nil {
		t.Fatal(err)
	}
	if event.Status != "started" || event.FinishedAt != nil || event.ContainsRawRows {
		t.Fatalf("unexpected initial model audit: %#v", event)
	}
	inputTokens, outputTokens, totalTokens := 100, 20, 120
	finished, err := service.FinishModelAudit(context.Background(), ModelAuditFinishInput{
		ID: event.ID, Status: "succeeded", InputTokens: &inputTokens,
		OutputTokens: &outputTokens, TotalTokens: &totalTokens, OutputBytes: 80,
	})
	if err != nil {
		t.Fatal(err)
	}
	if finished.Status != "succeeded" || finished.TotalTokens == nil || *finished.TotalTokens != 120 {
		t.Fatalf("model usage was not persisted: %#v", finished)
	}
	failure := "must not overwrite the terminal outcome"
	if _, err := service.FinishModelAudit(context.Background(), ModelAuditFinishInput{
		ID: event.ID, Status: "failed", OutputBytes: 0, Error: &failure,
	}); err == nil {
		t.Fatal("model audit terminal outcome was overwritten")
	}
	listed, err := service.ListModelAudits(context.Background())
	if err != nil || len(listed) != 1 || listed[0].PayloadSHA256 == "" {
		t.Fatalf("model audit was not listed: %#v, %v", listed, err)
	}
}

func TestModelAuditFailsClosedAndRecoversInterruptedRequests(t *testing.T) {
	service, dataset := importQueryFixture(t)
	invalid := datasetModelAuditInput(dataset)
	invalid.ContainsRawRows = true
	if _, err := service.StartModelAudit(context.Background(), invalid); err == nil {
		t.Fatal("raw-row model disclosure was accepted")
	}
	event, err := service.StartModelAudit(context.Background(), datasetModelAuditInput(dataset))
	if err != nil {
		t.Fatal(err)
	}
	if err := recoverInterruptedModelAudits(context.Background(), service.database); err != nil {
		t.Fatal(err)
	}
	recovered, err := service.getModelAudit(context.Background(), event.ID)
	if err != nil {
		t.Fatal(err)
	}
	if recovered.Status != "failed" || recovered.Error == nil || recovered.FinishedAt == nil {
		t.Fatalf("interrupted model audit was not closed: %#v", recovered)
	}
}
